### ISSUE TITLE
ProblemHttpResult doesn't show in generated OpenApi document

### ISSUE THREAD

**Issue by ilmax** [dotnet/aspnetcore#52424]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

It seems that `ProblemHttpResult` doesn't implement the `IEndpointMetadataProvider` interface hence, a method like the following:

```csharp
endpoints.MapGet("/", async Task<Results<Ok<GetClientsResponse>, ProblemHttpResult>>([FromServices] IRequiredActor<GetClientsActor> actor) =>
{
    var response = await /* Get clients from somewhere somehow.. Response is a poor man union type of IEnumerable<Client> | Error*/
      
    return response.Map<Results<Ok<GetClientsResponse>, ProblemHttpResult>>(
        clients => TypedResults.Ok(clients),
        error => TypedResults.Problem(detail: error.Reason, instance: error.Code, statusCode: StatusCodes.Status400BadRequest)
    );
});
```
Doesn't produce the expected OpenApi response types, only the 200 response ends up in the responses section of the document.

If I'm not reading the [documentation ](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/openapi?view=aspnetcore-8.0#set-responses-for-problemdetails) wrong, I'd expect the code above to produce 2 responses, a 200 and a 400. am I wrong?

### Expected Behavior

The problem details get's correctly added on the endpoint metadata when using `Result<Something, ProblemHttpResult>` and it shows up in the swagger UI

### .NET Version

net8.0

**Comment by captainsafia** [comment]
@ilmax Thanks for reporting this issue! As I recall, at the moment this is an intentional pattern. Unlike `ValidationProblemDetails`, which can default to a `400` status code. `ProblemDetails` does not have the same sane default. It could be `500`, but it could also be `401` or `505` or `429`.

In the past, we rejected an API proposal that attempted to solve this by introducing specific overloads for each status code (https://github.com/dotnet/aspnetcore/issues/47705) which is not ideal. The proposal indicates that there is a class of users interested in non-500 status codes.

I'm open to changing this given we get strong feedback that `500` is the correct default and won't cause surprises. I'll put this is the backlog and see if more folks upvote this.

**Comment by marinasundstrom** [comment]
I was surprised by this yesterday. Had to add .ProducesProblem(StatusCode…) for it to show up.

It makes sense since you can’t know what status code it is. Hence why you annotate.

But at least there could be a default for 500.

**Comment by linde12** [comment]
This is very confusing as a relatively green .NET developer. There are so many ways of doing the same thing. With builder pattern (.Produces), with annotations (ASP.NET MVC style?) and with generic return types (TypedResults).

I really liked the `TypedResults` concept where i could type whatever i return as generic arguments, e.g. `Results<Ok<UserResponse>, BadRequest<ErrorResponse>>` but when switching to `ProblemHttpResult` i now have to revert to the "old" and less cohesive (in my case, where each endpoint is a static method) way of annotating a response by imperatively adding `.ProducesProblem(...)` where i register the route (which is very easy to forget)

I expected the ProblemHttpResult to take a generic argument, just like e.g. BadRequest does for the body shape, but it does not. I mean like this:

```csharp
public static Results<Ok<UserCreatedResponse>, ProblemHttpResult<BadRequest>> CreateUser(...)
```

Again, new to C# and .NET world so i'm not sure if it is possible to have const generics in C#, but being able to do `ProblemHttpResult<400>` (or using a const integer instead of hardcoded 400) would be really nice.

It is not the end of the world, having to add `.ProducesProblem(400)`, but as i mentioned above some cohesion is lost and it is really easy to forget since it is not required/doesn't cause any errors if you forget to add it.

While we're at it, i'm also curious how we would type multiple kinds of errors (e.g. 400, 401, 500) using `TypedResults` - `Results` only takes 2 generic arguments (one for success and one for error) so how would i type a endpoint that returns 200, 400 and 402 with different bodies? Is the builder pattern or annotations the only escape hatch?

**Comment by alandraper** [comment]
> I'm open to changing this given we get strong feedback that `500` is the correct default and won't cause surprises. I'll put this is the backlog and see if more folks upvote this.

How do we upvote? This breaks the process for me.

**Comment by martincostello** [comment]
Add your 👍 to the issue description

**Comment by JayArrowz** [comment]
Noticed that this is one of the response types used by the `/login` endpoint for the package `Microsoft.AspNetCore.Identity` v10, in the extension function `IdentityApiEndpointRouteBuilderExtensions.MapIdentityApi<TUser>`

<img width="2106" height="1398" alt="Image" src="https://github.com/user-attachments/assets/e47ef11f-24bb-4e9c-9c1d-c2be4acb7541" />

The OpenApi generated for it had incomplete response types, it seems to completely omit the ProblemHttpResult type. I assume cause we don't know which status code is going to be returned.

```json
   "/login": {
      "post": {
        "tags": [
          "RentAudit"
        ],
        "parameters": [
          {
            "name": "useCookies",
            "in": "query",
            "schema": {
              "type": "boolean"
            }
          },
          {
            "name": "useSessionCookies",
            "in": "query",
            "schema": {
              "type": "boolean"
            }
          }
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/LoginRequest"
              }
            }
          },
          "required": true
        },
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/AccessTokenResponse"
                }
              }
            }
          }
        }
      }
    }
```

**Comment by emertechie** [comment]
@JayArrowz Yes I just ran into the same issue. I generate a TypeScript client from the spec, so this caused typing issues there with the missing 401 error response.

This is the patch I've implemented:

```cs
    app.MapIdentityApi<ApplicationUser>()
        .AddOpenApiOperationTransformer(async (operation, context, _) =>
        {
            // Transform /login endpoint to document the ProblemHttpResult 401 response which is
            // not included by default. See: https://github.com/dotnet/aspnetcore/issues/52424
            if (context.Description is { HttpMethod: "POST", RelativePath: "login" })
            {
                operation.Responses ??= new OpenApiResponses();
                operation.Responses.TryAdd("401", new OpenApiResponse
                {
                    Description = "Unauthorized",
                    Content = new Dictionary<string, OpenApiMediaType>
                    {
                        ["application/problem+json"] = new()
                        {
                            Schema = new OpenApiSchemaReference("ProblemDetails", context.Document)
                        }
                    }
                });
            }
        });
```

**Comment by PascalVorwerk** [comment]
Hi, which discussion or pull request made this ticket closed? Is there a fix/workaround for this issue? I am migrating my controller based api's to minimal api's and I am also trying to convert to using typedresults...

**Comment by admsteck** [comment]
I've run into this issue when trying to migrate to the TypedResults pattern in my controller-based API.  It also isn't limited to just `ProblemHttpResult` as it would also apply to `StatusCodeHttpResult`, `JsonHttpResult`, and others.  Adding `.ProducesProblem(StatusCode…)` as mentioned isn't an option for controller-based APIs.  

@captainsafia I would re-iterate @PascalVorwerk's question above.

**Comment by cremor** [comment]
> > > I'm open to changing this given we get strong feedback that `500` is the correct default and won't cause surprises. I'll put this is the backlog and see if more folks upvote this.
> >
> > How do we upvote? This breaks the process for me.
>
> Add your 👍 to the issue description

@martincostello I've upvoted the issue description because I agree that this is a problem. But I do **not** agree that 500 is the correct default status code to use here.

I usually don't document 5xx error codes in my OpenAPI document because the client shouldn't handle those specifically. Instead I use `[ProducesDefaultResponseType]` in my API controllers (I don't know if there is an equivalent for minimal APIs) to get a [Default Response](https://swagger.io/docs/specification/describing-responses/#default-response). Based on this I then document the status codes which the client has to handle.

> Adding `.ProducesProblem(StatusCode…)` as mentioned isn't an option for controller-based APIs.

@admsteck You can use the attribute `[ProducesResponseType(StatusCodes...)]` or `[ProducesResponseType<ProblemDetails>(StatusCodes...)]`
